# Spam Detector

In [1]:
# Import initial libraries and dependencies
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

### **Predictions**

Based on our experiences with the activities during Module 13, I would predict that the Random Forest Classifier model will perform better with this data. This is mostly because;

* In Activity 13.1.2, `LogisticRegression()` had an `accuracy_score` of **~95.9%**
* In Activity 13.2.6, `RandomForesetClassifier()` had an `accuracy_score` of **~96.7%**

Given the similarities with the dataset being used for this challenge, I'm expecting these two models to perform in line with previous experiences. Therefore, as stated, the Random Forest Classifier will have a slightly higher accuracy score.

## Split the Data into Training and Testing Sets

In [3]:
# Create the labels set `y` and features DataFrame `X`
y = data['spam']

X= data.copy().drop(columns='spam')

In [4]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [5]:
# Split the data into X_train, X_test, y_train, y_test
# (Note; `random_state=13` chosen to maintain consistent testing during development)
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=13)
X_train.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
2498            0.00                0.0           0.00           0.0   
1163            0.53                0.0           1.06           0.0   
2944            0.00                0.0           0.00           0.0   
1892            0.00                0.0           0.00           0.0   
4100            0.00                0.0           0.00           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
2498           0.00             0.0              0.00                 0.0   
1163           0.53             0.0              1.06                 0.0   
2944           0.00             0.0              0.00                 0.0   
1892           0.00             0.0              0.00                 0.0   
4100           0.00             0.0              0.00                 0.0   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
2498              0.0             0.0  ...                   0.0          0.0   
1163              0.0             0.0  ...                   0.0          0.0   
2944              0.0             0.0  ...                   0.0          0.0   
1892              0.0             0.0  ...                   0.0          0.0   
4100              0.0             0.0  ...                   0.0          0.0   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
2498        0.000        0.000        0.176        0.000          0.0   
1163        0.000        0.000        1.208        0.259          0.0   
2944        0.139        0.139        0.000        0.000          0.0   
1892        0.438        0.000        0.000        0.000          0.0   
4100        0.000        0.000        0.704        0.000          0.0   

      capital_run_length_average  capital_run_length_longest  \
2498                       1.758                           7   
1163                       5.558                          76   
2944                       1.763                           9   
1892                       2.448                          37   
4100                       1.000                           1   

      capital_run_length_total  
2498                        51  
1163                       189  
2944                        67  
1892                       120  
4100                        10  

[5 rows x 57 columns]

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [6]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scalar = StandardScaler()

In [7]:
# Fit the Standard Scaler with the training data
X_scalar = scalar.fit(X_train)

In [8]:
# Scale the training data
X_train_scaled = scalar.transform(X_train)
X_test_scaled = scalar.transform(X_test)

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [9]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression
lr_model = LogisticRegression(random_state=1).fit(X_train_scaled, y_train)

In [10]:
# Make and save testing predictions with the saved logistic regression model using the test data
lr_predictions = lr_model.predict(X_test_scaled)
# Review the predictions
lr_predictions

array([1, 0, 1, ..., 1, 1, 1], dtype=int64)

In [11]:
# Calculate the accuracy score by evaluating `y_test` vs. `lr_predictions`.
accuracy_score(y_test, lr_predictions)

0.9165942658557776

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [12]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
rf_model = RandomForestClassifier(random_state=1).fit(X_train_scaled, y_train)

In [13]:
# Make and save testing predictions with the saved logistic regression model using the test data
rf_predictions = rf_model.predict(X_test_scaled)

# Review the predictions
rf_predictions

array([1, 0, 0, ..., 1, 1, 1], dtype=int64)

In [14]:
# Calculate the accuracy score by evaluating `y_test` vs. `rf_predictions`.
accuracy_score(y_test, rf_predictions)

0.9513466550825369

## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

### **Evaluation**

After running the dataset through the two models, I was able to confirm the accuracy scores of both models;

* `LogisticRegression()` had an `accuracy_score` of **~91.7%**
* `RandomForesetClassifier()` had an `accuracy_score` of **~95.1%**

Clearly the better performing model for this datset was the Random Forest Classifier. These findings support my earlier prediction and prove my assumption correct.

## Additional Confirmtion

Below I will test the precision scores for both models to further support my earlier predictions.

In [15]:
# Import `precision_score()`
from sklearn.metrics import precision_score

In [16]:
# Calculate `precision_score()` for `y_test` vs. both `lr_predictions` and `rf_predictions`
lr_precision = precision_score(y_test, lr_predictions)
rf_precision = precision_score(y_test, rf_predictions)

In [17]:
# Display both precision scores
print(f'LR precision: {lr_precision}')
print(f'RF precision: {rf_precision}')

LR precision: 0.9080962800875274
RF precision: 0.9578713968957872


### **Final Confirmation**

This additional step found the precision scores of both models;

* `LogisticRegression()` had an `precision_score` of **~90.8%**
* `RandomForesetClassifier()` had an `precision_score` of **~95.8%**

These findings further confrim my earliest assumptions by showing the Random Forest Classifier model to have the higher accruacy *and* precision scores.